### PR TITLE
Upgrade Channel resource to API Platform 3

### PR DESCRIPTION
**Pull request by TheMilek**
| Q               | A                                                            |
|-----------------|--------------------------------------------------------------|
| Branch?         | api-platform-3                   |
| Bug fix?        | no                                                       |
| New feature?    | yes                                                       |
| BC breaks?      | no                                                      |
| License         | MIT                                                          |

### CODE REVIEW

File: src/Sylius/Bundle/ApiBundle/Resources/config/api_platform/resources/Channel.xml [hunk 1]
```diff
@@ -16,7 +16,7 @@
     xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
     xsi:schemaLocation="https://api-platform.com/schema/metadata/resources-3.0 https://api-platform.com/schema/metadata/resources-3.0.xsd"
 >
-    <resource class="Sylius\Component\Core\Model\Channel" shortName="Channel">
+    <resource class="Sylius\Component\Core\Model\Channel" processor="Sylius\Bundle\ApiBundle\StateProcessor\ChannelProcessor" shortName="Channel">
```
Comment (Wojdylak) [inline src/Sylius/Bundle/ApiBundle/Resources/config/api_platform/resources/Channel.xml]:
```suggestion
    <resource class="Sylius\Component\Core\Model\Channel" shortName="Channel" processor="Sylius\Bundle\ApiBundle\StateProcessor\ChannelProcessor">
```
For me that is more readable, WDYT?

File: src/Sylius/Bundle/ApiBundle/Resources/config/api_platform/resources/ShopBillingData.xml [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+                <value name="groups">
+                    <values>
+                        <value>sylius</value>
+                    </values>
+                </value>
+            </values>
+        </validationContext>
+
+        <operations>
+            <operation class="ApiPlatform\Metadata\Get" routePrefix="admin" >
```
Comment (Wojdylak) [inline src/Sylius/Bundle/ApiBundle/Resources/config/api_platform/resources/ShopBillingData.xml]: You use `uriTemplate` once and `routePrefix` another time, why?

File: src/Sylius/Bundle/ApiBundle/StateProcessor/ChannelProcessor.php [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+use ApiPlatform\Metadata\Operation;
+use ApiPlatform\State\ProcessorInterface;
+use Sylius\Bundle\ApiBundle\Exception\ChannelCannotBeRemoved;
+use Sylius\Component\Channel\Checker\ChannelDeletionCheckerInterface;
+use Sylius\Component\Core\Model\ChannelInterface;
+use Symfony\Component\HttpFoundation\Response;
+use Webmozart\Assert\Assert;
+
+/**
+ * @experimental
```
Comment (Wojdylak) [inline src/Sylius/Bundle/ApiBundle/StateProcessor/ChannelProcessor.php]: Not anymore :sunglasses:

File: src/Sylius/Bundle/ApiBundle/StateProcessor/ChannelProcessor.php [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+{
+    public function __construct(
+        private ProcessorInterface $persistProcessor,
+        private ProcessorInterface $removeProcessor,
+        private ChannelDeletionCheckerInterface $channelDeletionChecker,
+    ) {
+    }
+
+    /**
+     * @param ChannelInterface $data
```
Comment (Wojdylak) [inline src/Sylius/Bundle/ApiBundle/StateProcessor/ChannelProcessor.php]: Is it necessary if there is `assert` in first line of the method?

File: tests/Api/Admin/ChannelsTest.php [hunk 5]
```diff
@@ -158,49 +159,51 @@ public function it_updates_an_existing_channel(): void
...
-            server: $header,
-        );
-
-        $this->assertResponse(
-            $this->client->getResponse(),
-            'admin/shop_billing_data/put_shop_billing_data_response',
-            Response::HTTP_OK,
-        );
-    }
+    // todo: needs ShopBillingData resource to be fully implemented
```
Comment (Wojdylak) [inline tests/Api/Admin/ChannelsTest.php:162]: Could you add a note about this to the ShopBillingData task?

File: src/Sylius/Bundle/ApiBundle/spec/StateProcessor/ChannelProcessorSpec.php [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+{
+    function let(
+        ProcessorInterface $persistProcessor,
+        ProcessorInterface $removeProcessor,
+        ChannelDeletionCheckerInterface $channelDeletionChecker,
+    ): void {
+        $this->beConstructedWith($persistProcessor, $removeProcessor, $channelDeletionChecker);
+    }
+
+    function it_throws_an_exception_if_object_is_not_a_country(
```
Comment (Wojdylak) [inline src/Sylius/Bundle/ApiBundle/spec/StateProcessor/ChannelProcessorSpec.php]:
```suggestion
    function it_throws_an_exception_if_object_is_not_a_channel(
```

File: src/Sylius/Bundle/ApiBundle/StateProvider/Shop/ChannelProvider.php [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+        if ($this->isShopGetCollectionOperation($operation, $context) && isset($context['sylius_api_channel'])) {
+            return [$context['sylius_api_channel']];
+        }
+
+        throw new \RuntimeException('Only Shop GET collection operation is supported.');
+    }
+
+    private function isShopGetCollectionOperation(Operation $operation, array $context): bool
+    {
+        return $operation instanceof GetCollection && \str_starts_with($context['operation_name'], '_api_/shop');
```
Comment (Wojdylak) [inline src/Sylius/Bundle/ApiBundle/StateProvider/Shop/ChannelProvider.php]: There's a `SectionProvider`. Perhaps we should use it, but I'm unsure if it still works with API Platform 3."

File: src/Sylius/Bundle/ApiBundle/Resources/config/services/state_providers.xml [hunk 8]
```diff
@@ -27,5 +27,10 @@
         <service id="Sylius\Bundle\ApiBundle\StateProvider\Admin\ResetPasswordProvider">
             <tag name="api_platform.state_provider" priority="10" />
         </service>
+
+        <service id="Sylius\Bundle\ApiBundle\StateProvider\Shop\ChannelProvider">
```
Comment (Wojdylak) [inline src/Sylius/Bundle/ApiBundle/Resources/config/services/state_providers.xml]:
```suggestion
        <service id="sylius.api.state_provider.shop.channel" class="Sylius\Bundle\ApiBundle\StateProvider\Shop\ChannelProvider">
```
#15883
And to your other services

**Review by Wojdylak — APPROVED** [review]

**Review by NoResponseMate — APPROVED** [review]

File: src/Sylius/Bundle/ApiBundle/Resources/config/services/state_providers.xml (hunk 8 above)
Comment (NoResponseMate) [reply src/Sylius/Bundle/ApiBundle/Resources/config/services/state_providers.xml]: Basing on that convention, it should start with `sylius_api.`

**Comment by NoResponseMate** [comment]
Thank you, @TheMilek!

**Pull request merged**